File: backend/src/execution/simulatedExecutor.js
/**
 * simulatedExecutor.js — Handles execution for the full batch (40-100 cases).
 *
 * Never calls the real Razorpay API. Uses seeded PRNG for deterministic,
 * reproducible outcomes — the same case always resolves the same way.
 * ~65% of payment links resolve to "paid", remainder unresolved/expired.
 */

const { createSeededRng } = require('../utils/seededRandom');

/**
 * Execute a recovery action in simulation mode.
 *
 * @param {object} caseData - The case document from MongoDB
 * @param {object} plan - The approved recovery plan
 * @returns {object} Updated execution fields + status to set on the case
 */
function executeSimulated(caseData, plan) {
  const rng = createSeededRng(caseData._id.toString());
  const now = new Date();
  const recommendation = plan.recommendation;

  // Actions that involve a payment link
  if (recommendation === 'CREATE_PAYMENT_LINK' || recommendation === 'OFFER_DISCOUNT') {
    const simReference = `sim_plink_${caseData._id.toString().slice(-8)}`;
    const paymentOutcome = rng(); // deterministic [0, 1)

    let paymentLinkStatus;
    let status;

    if (paymentOutcome < 0.65) {
      // ~65% chance: customer pays
      paymentLinkStatus = 'paid';
      status = 'recovered';
    } else if (paymentOutcome < 0.85) {
      // ~20% chance: link expires (customer didn't pay)
      paymentLinkStatus = 'expired';
      status = 'unrecovered_expired';
    } else {
      // ~15% chance: still outstanding
      paymentLinkStatus = 'created';
      status = 'link_created';
    }

    return {
      execution: {
        executor_type: 'simulated',
        action_taken: recommendation,
        result: 'success',
        razorpay_reference: simReference,
        payment_link_status: paymentLinkStatus,
        timestamp: now
      },
      status,
      // Increment contact_count for payment link actions
      contact_count: caseData.contact_count + 1,
      last_contacted_at: now
    };
  }

  // SEND_REMINDER — simulate message sent
  if (recommendation === 'SEND_REMINDER') {
    return {
      execution: {
        executor_type: 'simulated',
        action_taken: 'SEND_REMINDER',
        result: 'success',
        razorpay_reference: null,
        payment_link_status: null,
        timestamp: now
      },
      status: 'link_created', // reminder sent, waiting for customer action
      contact_count: caseData.contact_count + 1,
      last_contacted_at: now
    };
  }

  // ESCALATE_TO_HUMAN / DO_NOT_CONTACT — no external action
  return {
    execution: {
      executor_type: 'simulated',
      action_taken: recommendation,
      result: 'not_attempted',
      razorpay_reference: null,
      payment_link_status: null,
      timestamp: now
    },
    status: recommendation === 'DO_NOT_CONTACT' ? 'stopped_safely' : 'needs_merchant_approval'
  };
}

module.exports = { executeSimulated };

File: backend/app/execution/simulated_executor.py
from datetime import datetime, timezone
from app.models.case import Case, Plan, Execution, ExecutorType, ExecutionResult, PaymentLinkStatus, CaseStatus
from app.utils.seeded_random import create_seeded_rng

def execute_simulated(case_data: Case, plan: Plan) -> dict:
    rng = create_seeded_rng(str(case_data.id))
    now = datetime.now(timezone.utc)
    recommendation = plan.recommendation.value if plan.recommendation else None

    if recommendation in ['CREATE_PAYMENT_LINK', 'OFFER_DISCOUNT']:
        sim_reference = f"sim_plink_{str(case_data.id)[-8:]}"
        payment_outcome = rng()

        payment_link_status = None
        status = None

        if payment_outcome < 0.65:
            payment_link_status = PaymentLinkStatus.PAID
            status = CaseStatus.RECOVERED
        elif payment_outcome < 0.85:
            payment_link_status = PaymentLinkStatus.EXPIRED
            status = CaseStatus.UNRECOVERED_EXPIRED
        else:
            payment_link_status = PaymentLinkStatus.CREATED
            status = CaseStatus.LINK_CREATED

        return {
            "execution": Execution(
                executor_type=ExecutorType.SIMULATED,
                action_taken=recommendation,
                result=ExecutionResult.SUCCESS,
                razorpay_reference=sim_reference,
                payment_link_status=payment_link_status,
                timestamp=now
            ),
            "status": status,
            "contact_count": case_data.contact_count + 1,
            "last_contacted_at": now
        }

    if recommendation == 'SEND_REMINDER':
        return {
            "execution": Execution(
                executor_type=ExecutorType.SIMULATED,
                action_taken='SEND_REMINDER',
                result=ExecutionResult.SUCCESS,
                razorpay_reference=None,
                payment_link_status=None,
                timestamp=now
            ),
            "status": CaseStatus.LINK_CREATED,
            "contact_count": case_data.contact_count + 1,
            "last_contacted_at": now
        }

    return {
        "execution": Execution(
            executor_type=ExecutorType.SIMULATED,
            action_taken=recommendation,
            result=ExecutionResult.NOT_ATTEMPTED,
            razorpay_reference=None,
            payment_link_status=None,
            timestamp=now
        ),
        "status": CaseStatus.STOPPED_SAFELY if recommendation == 'DO_NOT_CONTACT' else CaseStatus.NEEDS_MERCHANT_APPROVAL
    }
